File: backend/src/algorithms/slidingWindow.ts
import { SlidingWindowEvaluationResult } from './types.js';

export interface SlidingWindowParams {
  timestamps: (Date | number)[];
  limit: number;
  windowSize: number; // in seconds
}

/**
 * Pure Sliding Window algorithm implementation.
 *
 * Properties:
 * - Maintains a log of timestamps within the trailing window `[now - windowSize, now]`.
 * - Expires and removes timestamps older than `now - windowSize`.
 * - If remaining active timestamp count < limit: ALLOW and record timestamp.
 * - If count >= limit: DENY with retryAfter computed from the earliest timestamp in the window.
 */
export function evaluateSlidingWindow(
  params: SlidingWindowParams,
  now: Date | number = Date.now()
): SlidingWindowEvaluationResult {
  const { limit, windowSize } = params;

  if (limit <= 0) {
    throw new Error('limit must be greater than 0');
  }
  if (windowSize <= 0) {
    throw new Error('windowSize must be greater than 0');
  }

  const nowMs = typeof now === 'number' ? now : now.getTime();
  const windowSizeMs = windowSize * 1000;
  const windowStartMs = nowMs - windowSizeMs;

  // Filter timestamps within [now - windowSize, now]
  const validTimestamps: Date[] = [];
  for (const t of params.timestamps) {
    const tMs = typeof t === 'number' ? t : t.getTime();
    if (tMs >= windowStartMs && tMs <= nowMs) {
      validTimestamps.push(typeof t === 'number' ? new Date(t) : t);
    }
  }

  // Sort chronologically ascending
  validTimestamps.sort((a, b) => a.getTime() - b.getTime());

  const currentCount = validTimestamps.length;
  const nowSec = Math.floor(nowMs / 1000);

  if (currentCount < limit) {
    const remaining = limit - currentCount - 1;
    
    // Earliest expiry timestamp or end of current window
    const earliestMs = validTimestamps.length > 0 ? validTimestamps[0].getTime() : nowMs;
    const reset = Math.max(nowSec, Math.ceil((earliestMs + windowSizeMs) / 1000));

    return {
      decision: 'ALLOW',
      remaining: Math.max(0, remaining),
      limit: Math.floor(limit),
      reset,
      validTimestamps: [...validTimestamps, new Date(nowMs)],
    };
  } else {
    // DENY: Window is saturated
    const earliestMs = validTimestamps[0].getTime();
    const msUntilOldestExpires = Math.max(0, earliestMs + windowSizeMs - nowMs);
    const retryAfter = Math.max(1, Math.ceil(msUntilOldestExpires / 1000));
    const reset = Math.max(nowSec, Math.ceil((earliestMs + windowSizeMs) / 1000));

    return {
      decision: 'DENY',
      remaining: 0,
      limit: Math.floor(limit),
      reset,
      retryAfter,
      validTimestamps,
    };
  }
}

File: backend/src/repositories/lua/slidingWindow.lua
-- Redis Lua Script: Atomic Sliding Window Evaluation
-- KEYS[1]: client:{clientKey}:window (Sorted Set)
-- ARGV[1]: limit (maximum allowed requests)
-- ARGV[2]: windowSizeSec (window size in seconds)
-- ARGV[3]: nowMs (current Unix epoch in milliseconds)
-- ARGV[4]: requestId (unique identifier for this request member)

local windowKey = KEYS[1]
local limit = tonumber(ARGV[1])
local windowSizeSec = tonumber(ARGV[2])
local nowMs = tonumber(ARGV[3])
local requestId = ARGV[4]

local windowSizeMs = windowSizeSec * 1000.0
local windowStartMs = nowMs - windowSizeMs

-- 1. Prune expired entries older than now - windowSize
redis.call('ZREMRANGEBYSCORE', windowKey, '-inf', '(' .. tostring(windowStartMs))

-- 2. Count active entries in current window
local currentCount = redis.call('ZCARD', windowKey)
local nowSec = math.floor(nowMs / 1000.0)

if currentCount < limit then
  -- ALLOW: Record request timestamp
  redis.call('ZADD', windowKey, nowMs, requestId)
  
  -- Set TTL to 2x window size
  redis.call('EXPIRE', windowKey, math.ceil(windowSizeSec * 2 + 60))

  local remaining = math.max(0, limit - currentCount - 1)
  local reset = math.max(nowSec, math.ceil((nowMs + windowSizeMs) / 1000.0))

  -- Return { is_allowed (1/0), remaining, limit, reset, retry_after }
  return { 1, remaining, limit, reset, 0 }
else
  -- DENY: Quota exceeded
  local oldest = redis.call('ZRANGE', windowKey, 0, 0, 'WITHSCORES')
  local oldestMs = (oldest and oldest[2]) and tonumber(oldest[2]) or nowMs
  local msUntilExpiry = math.max(0, oldestMs + windowSizeMs - nowMs)
  local retryAfter = math.max(1, math.ceil(msUntilExpiry / 1000.0))
  local reset = math.max(nowSec, math.ceil((oldestMs + windowSizeMs) / 1000.0))

  return { 0, 0, limit, reset, retryAfter }
end
